src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -759,6 +639,9 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
                         {
                             if (!_sslAuthenticationOptions!.AllowRenegotiation)
                             {
+                                _handshakeWaiter!.SetResult(false);
+                                _handshakeWaiter = null;
```

Comment:
It'd be helpful to add a comment somewhere detailing the locking scheme being used.  That will, for example, help to explain why it's ok for this nulling out of the field to be done (e.g. because there's only ever one reader at a time, and the reader is responsible for initiating all renegotations, and thus for creating, completing, and removing a waiter).

Reply:
updated. I added long comment to where the waiter is created and the logic starts.